neqo-transport/src/shuffle.rs

```diff
@@ -0,0 +1,246 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::{collections::BinaryHeap, mem, ops::Range};
+
+use neqo_crypto::random;
+
+/// [Fisher–Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Modern_method)
+/// Modified to make sure no element stays in place if `a` has more than one element.
+///
+/// We could use <https://docs.rs/rand/latest/rand/seq/trait.SliceRandom.html#tymethod.shuffle>
+/// instead, but we're currently not depending on the `rand` crate and this is a simple enough
+/// so we don't need to.
+fn shuffle<T>(a: &mut [T]) {
+    // To shuffle an array a of n elements (indices 0..n-1)
+    const USIZE: usize = mem::size_of::<usize>();
+    let n: usize = a.len();
+    if n < 2 {
+        return;
+    }
+    for i in 0..(n - 1) {
+        // j ← random integer such that i ≤ j < n
+        let j = usize::from_ne_bytes(random::<USIZE>()) % (n - i - 1) + i + 1;
+        // Exchange a[i] and a[j]
+        debug_assert!(i != j);
+        a.swap(i, j);
+    }
+}
+
+/// Find the ranges of all sequences of two or more ASCII "LDHD" (letters, digits, hyphens, dots)
+/// bytes in `data`, and return the `n` longest ones in ascending order by start index.
+fn ascii_sequences(data: &[u8], n: usize) -> impl Iterator<Item = Range<usize>> {
```

Comment:
I'm still not that enthusiastic about this approach, but here's a way to improve it some.

Change the signature to `fn ascii_sequences<const N: usize>(data: &[u8]) -> impl Interator...`.

Then collect the value into an array (`[_;  N]`) for sorting rather than a Vec.  You might have to unwrap, but that's not so bad.

Reply:
I did this in e9f3b58597d6dd202e7a7cb7ea7100f1ec51b3cf.

Reply:
But I don't think I can collect into an array, because there might be fewer than `N` ranges. I *could* collect into an array with `Option<Range<usize>>` and have some of them be `None` in that case - is that worth it?